superset/assets/src/middleware/loggerMiddleware.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/* eslint-disable camelcase */
+/* eslint prefer-const: 2 */
+import { debounce } from 'lodash';
+import shortid from 'shortid';
+import { SupersetClient } from '@superset-ui/connection';
+
+import { LOG_EVENT } from '../logger/actions';
+import { LOG_EVENT_TYPE_TIMING } from '../logger/LogUtils';
+
+const LOG_ENDPOINT = '/superset/log/?explode=events';
+
+const sendBeacon = debounce(({ eventDataList = [] }) => {
+  let length = eventDataList.length;
+  if (length <= 0) {
+    return;
+  }
+
+  let endpoint = LOG_ENDPOINT;
+  const { source, source_id } = eventDataList[0];
+  // backend logs treat these request params as first-class citizens
+  if (source === 'dashboard') {
```

Comment:
What if there are multiple `events` in `eventDataList` and they have different `source`?

Reply:
you are right!  all root component share same middleware.